File: backend/src/main/java/com/society/module/owner/service/OwnerService.java
package com.society.module.owner.service;

import com.society.common.PagedResponse;
import com.society.enums.OccupancyStatus;
import com.society.enums.OwnerStatus;
import com.society.exception.BusinessException;
import com.society.exception.ResourceNotFoundException;
import com.society.module.owner.dto.*;
import com.society.module.owner.entity.Owner;
import com.society.module.owner.entity.OwnershipHistory;
import com.society.module.owner.entity.Unit;
import com.society.module.owner.entity.UnitOwner;
import com.society.module.owner.repository.OwnerRepository;
import com.society.module.owner.repository.OwnershipHistoryRepository;
import com.society.module.owner.repository.UnitOwnerRepository;
import com.society.module.owner.repository.UnitRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class OwnerService {

    private final OwnerRepository ownerRepository;
    private final UnitRepository unitRepository;
    private final UnitOwnerRepository unitOwnerRepository;
    private final OwnershipHistoryRepository ownershipHistoryRepository;

    // ==================== OWNER CRUD ====================

    @Transactional
    public OwnerDTO createOwner(OwnerCreateRequest request) {
        Owner owner = Owner.builder()
                .fullName(request.getFullName())
                .contactNumber(request.getContactNumber())
                .alternateNumber(request.getAlternateNumber())
                .email(request.getEmail())
                .aadharNumber(request.getAadharNumber())
                .panNumber(request.getPanNumber())
                .permanentAddress(request.getPermanentAddress())
                .occupation(request.getOccupation())
                .emergencyContactName(request.getEmergencyContactName())
                .emergencyContactPhone(request.getEmergencyContactPhone())
                .status(OwnerStatus.ACTIVE)
                .build();

        owner = ownerRepository.save(owner);
        return mapToDTO(owner);
    }

    @Transactional
    public OwnerDTO updateOwner(Long ownerId, OwnerUpdateRequest request) {
        Owner owner = ownerRepository.findById(ownerId)
                .orElseThrow(() -> new ResourceNotFoundException("Owner", "ownerId", ownerId));

        owner.setFullName(request.getFullName());
        owner.setContactNumber(request.getContactNumber());
        owner.setAlternateNumber(request.getAlternateNumber());
        owner.setEmail(request.getEmail());
        owner.setAadharNumber(request.getAadharNumber());
        owner.setPanNumber(request.getPanNumber());
        owner.setPermanentAddress(request.getPermanentAddress());
        owner.setOccupation(request.getOccupation());
        owner.setEmergencyContactName(request.getEmergencyContactName());
        owner.setEmergencyContactPhone(request.getEmergencyContactPhone());

        owner = ownerRepository.save(owner);
        return mapToDTO(owner);
    }

    public OwnerDTO getOwnerById(Long ownerId) {
        Owner owner = ownerRepository.findById(ownerId)
                .orElseThrow(() -> new ResourceNotFoundException("Owner", "ownerId", ownerId));
        return mapToDTO(owner);
    }

    public PagedResponse<OwnerDTO> getAllOwners(int page, int size, String status, String search) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("ownerId").ascending());
        Page<Owner> ownerPage;

        if (search != null && !search.isBlank()) {
            if (status != null && !status.isBlank()) {
                ownerPage = ownerRepository.searchOwners(OwnerStatus.valueOf(status), search, pageable);
            } else {
                ownerPage = ownerRepository.searchAllOwners(search, pageable);
            }
        } else if (status != null && !status.isBlank()) {
            ownerPage = ownerRepository.findByStatus(OwnerStatus.valueOf(status), pageable);
        } else {
            ownerPage = ownerRepository.findAll(pageable);
        }

        List<OwnerDTO> content = ownerPage.getContent().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());

        return PagedResponse.<OwnerDTO>builder()
                .content(content)
                .page(ownerPage.getNumber())
                .size(ownerPage.getSize())
                .totalElements(ownerPage.getTotalElements())
                .totalPages(ownerPage.getTotalPages())
                .last(ownerPage.isLast())
                .build();
    }

    public List<OwnerDTO> getActiveOwnersList() {
        return ownerRepository.findByStatusOrderByFullNameAsc(OwnerStatus.ACTIVE)
                .stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    // ==================== OWNERSHIP TRANSFER ====================

    @Transactional
    public OwnershipHistoryDTO transferOwnership(OwnershipTransferRequest request) {
        Unit unit = unitRepository.findById(request.getUnitId())
                .orElseThrow(() -> new ResourceNotFoundException("Unit", "unitId", request.getUnitId()));

        Owner newOwner = ownerRepository.findById(request.getNewOwnerId())
                .orElseThrow(() -> new ResourceNotFoundException("Owner", "ownerId", request.getNewOwnerId()));

        if (newOwner.getStatus() != OwnerStatus.ACTIVE) {
            throw new BusinessException("New owner must have ACTIVE status for transfer");
        }

        // Close current ownership history record
        OwnershipHistory currentHistory = ownershipHistoryRepository
                .findCurrentOwnershipByUnitId(unit.getUnitId())
                .orElse(null);

        if (currentHistory != null) {
            currentHistory.setOwnershipEndDate(request.getTransferDate());
            ownershipHistoryRepository.save(currentHistory);

            // Mark old owner as TRANSFERRED if they don't own any other units
            Owner oldOwner = currentHistory.getOwner();
            List<Unit> otherUnits = unitRepository.findByOwnerId(oldOwner.getOwnerId());
            long otherUnitCount = otherUnits.stream()
                    .filter(u -> !u.getUnitId().equals(unit.getUnitId()))
                    .count();
            if (otherUnitCount == 0) {
                oldOwner.setStatus(OwnerStatus.TRANSFERRED);
                ownerRepository.save(oldOwner);
            }
        }

        // Remove all existing owners from the unit (transfer clears all co-owners)
        List<UnitOwner> existingOwners = unitOwnerRepository.findByUnit_UnitId(unit.getUnitId());
        unitOwnerRepository.deleteAll(existingOwners);

        // Add new owner as primary with 100% ownership
        UnitOwner newUnitOwner = UnitOwner.builder()
                .unit(unit)
                .owner(newOwner)
                .isPrimary(true)
                .ownershipPercentage(new java.math.BigDecimal("100.00"))
                .addedOn(LocalDateTime.now())
                .addedBy("SYSTEM")
                .build();
        unitOwnerRepository.save(newUnitOwner);

        // Update unit occupancy
        unit.setOccupancyStatus(OccupancyStatus.SELF_OCCUPIED);
        unitRepository.save(unit);

        // Create new ownership history record
        OwnershipHistory newHistory = OwnershipHistory.builder()
                .unit(unit)
                .owner(newOwner)
                .ownershipStartDate(request.getTransferDate())
                .ownershipEndDate(null)
                .transferType(request.getTransferType())
                .remarks(request.getRemarks())
                .recordedBy("SYSTEM")
                .recordedOn(LocalDateTime.now())
                .build();
        newHistory = ownershipHistoryRepository.save(newHistory);

        return mapToHistoryDTO(newHistory);
    }

    // ==================== OWNERSHIP HISTORY ====================

    public List<OwnershipHistoryDTO> getOwnershipHistoryByUnit(Long unitId) {
        unitRepository.findById(unitId)
                .orElseThrow(() -> new ResourceNotFoundException("Unit", "unitId", unitId));

        return ownershipHistoryRepository.findByUnitIdOrderByStartDateDesc(unitId)
                .stream()
                .map(this::mapToHistoryDTO)
                .collect(Collectors.toList());
    }

    public List<OwnershipHistoryDTO> getOwnershipHistoryByOwner(Long ownerId) {
        ownerRepository.findById(ownerId)
                .orElseThrow(() -> new ResourceNotFoundException("Owner", "ownerId", ownerId));

        return ownershipHistoryRepository.findByOwnerIdOrderByStartDateDesc(ownerId)
                .stream()
                .map(this::mapToHistoryDTO)
                .collect(Collectors.toList());
    }

    // ==================== MAPPERS ====================

    private OwnerDTO mapToDTO(Owner owner) {
        // Fetch unit numbers owned by this owner
        String unitNumbers = unitOwnerRepository.findByOwner_OwnerId(owner.getOwnerId())
                .stream()
                .map(uo -> uo.getUnit().getUnitNumber())
                .collect(Collectors.joining(", "));

        return OwnerDTO.builder()
                .ownerId(owner.getOwnerId())
                .fullName(owner.getFullName())
                .contactNumber(owner.getContactNumber())
                .alternateNumber(owner.getAlternateNumber())
                .email(owner.getEmail())
                .aadharNumber(owner.getAadharNumber())
                .panNumber(owner.getPanNumber())
                .permanentAddress(owner.getPermanentAddress())
                .occupation(owner.getOccupation())
                .photoPath(owner.getPhotoPath())
                .emergencyContactName(owner.getEmergencyContactName())
                .emergencyContactPhone(owner.getEmergencyContactPhone())
                .status(owner.getStatus())
                .unitNumbers(unitNumbers)
                .createdBy(owner.getCreatedBy())
                .createdOn(owner.getCreatedOn())
                .modifiedBy(owner.getModifiedBy())
                .modifiedOn(owner.getModifiedOn())
                .build();
    }

    private OwnershipHistoryDTO mapToHistoryDTO(OwnershipHistory history) {
        return OwnershipHistoryDTO.builder()
                .historyId(history.getHistoryId())
                .unitId(history.getUnit().getUnitId())
                .unitNumber(history.getUnit().getUnitNumber())
                .ownerId(history.getOwner().getOwnerId())
                .ownerName(history.getOwner().getFullName())
                .ownershipStartDate(history.getOwnershipStartDate())
                .ownershipEndDate(history.getOwnershipEndDate())
                .transferType(history.getTransferType())
                .transferDocumentPath(history.getTransferDocumentPath())
                .remarks(history.getRemarks())
                .recordedBy(history.getRecordedBy())
                .recordedOn(history.getRecordedOn())
                .build();
    }
}

File: frontend/src/app/core/services/owner.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '@env/environment';
import { ApiResponse, PagedResponse } from '../models/api-response.model';
import { Owner, OwnerCreateRequest, OwnerUpdateRequest, Unit, UnitCreateRequest, UnitOwner, AddCoOwnerRequest, OwnershipHistory, OwnershipTransferRequest } from '../models/owner.model';

@Injectable({ providedIn: 'root' })
export class OwnerService {

  private readonly apiUrl = `${environment.apiUrl}/owners`;
  private readonly unitUrl = `${environment.apiUrl}/units`;

  constructor(private http: HttpClient) {}

  // ===== OWNERS =====
  createOwner(request: OwnerCreateRequest): Observable<ApiResponse<Owner>> {
    return this.http.post<ApiResponse<Owner>>(this.apiUrl, request);
  }

  updateOwner(ownerId: number, request: OwnerUpdateRequest): Observable<ApiResponse<Owner>> {
    return this.http.put<ApiResponse<Owner>>(`${this.apiUrl}/${ownerId}`, request);
  }

  getOwnerById(ownerId: number): Observable<ApiResponse<Owner>> {
    return this.http.get<ApiResponse<Owner>>(`${this.apiUrl}/${ownerId}`);
  }

  getAllOwners(page = 0, size = 20, status?: string, search?: string): Observable<ApiResponse<PagedResponse<Owner>>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (status) params = params.set('status', status);
    if (search) params = params.set('search', search);
    return this.http.get<ApiResponse<PagedResponse<Owner>>>(this.apiUrl, { params });
  }

  getActiveOwnersList(): Observable<ApiResponse<Owner[]>> {
    return this.http.get<ApiResponse<Owner[]>>(`${this.apiUrl}/active-list`);
  }

  transferOwnership(request: OwnershipTransferRequest): Observable<ApiResponse<OwnershipHistory>> {
    return this.http.post<ApiResponse<OwnershipHistory>>(`${this.apiUrl}/transfer`, request);
  }

  getHistoryByUnit(unitId: number): Observable<ApiResponse<OwnershipHistory[]>> {
    return this.http.get<ApiResponse<OwnershipHistory[]>>(`${this.apiUrl}/history/unit/${unitId}`);
  }

  getHistoryByOwner(ownerId: number): Observable<ApiResponse<OwnershipHistory[]>> {
    return this.http.get<ApiResponse<OwnershipHistory[]>>(`${this.apiUrl}/history/owner/${ownerId}`);
  }

  // ===== UNITS =====
  createUnit(request: UnitCreateRequest): Observable<ApiResponse<Unit>> {
    return this.http.post<ApiResponse<Unit>>(this.unitUrl, request);
  }

  updateUnit(unitId: number, request: UnitCreateRequest): Observable<ApiResponse<Unit>> {
    return this.http.put<ApiResponse<Unit>>(`${this.unitUrl}/${unitId}`, request);
  }

  getUnitById(unitId: number): Observable<ApiResponse<Unit>> {
    return this.http.get<ApiResponse<Unit>>(`${this.unitUrl}/${unitId}`);
  }

  getAllUnits(page = 0, size = 20, wing?: string, unitType?: string, occupancyStatus?: string): Observable<ApiResponse<PagedResponse<Unit>>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (wing) params = params.set('wing', wing);
    if (unitType) params = params.set('unitType', unitType);
    if (occupancyStatus) params = params.set('occupancyStatus', occupancyStatus);
    return this.http.get<ApiResponse<PagedResponse<Unit>>>(this.unitUrl, { params });
  }

  getUnitsByOwner(ownerId: number): Observable<ApiResponse<Unit[]>> {
    return this.http.get<ApiResponse<Unit[]>>(`${this.unitUrl}/by-owner/${ownerId}`);
  }

  getOccupancySummary(): Observable<ApiResponse<any>> {
    return this.http.get<ApiResponse<any>>(`${this.unitUrl}/summary`);
  }

  // ===== CO-OWNER MANAGEMENT =====
  addOwnerToUnit(request: AddCoOwnerRequest): Observable<ApiResponse<Unit>> {
    return this.http.post<ApiResponse<Unit>>(`${this.unitUrl}/owners`, request);
  }

  removeOwnerFromUnit(unitId: number, ownerId: number): Observable<ApiResponse<Unit>> {
    return this.http.delete<ApiResponse<Unit>>(`${this.unitUrl}/${unitId}/owners/${ownerId}`);
  }

  getUnitOwners(unitId: number): Observable<ApiResponse<UnitOwner[]>> {
    return this.http.get<ApiResponse<UnitOwner[]>>(`${this.unitUrl}/${unitId}/owners`);
  }
}
